hubblestack/utils/signing.py

```diff
@@ -0,0 +1,474 @@
+    """
+    log.debug("verify_files(%s, mfname=%s, sfname=%s, public_crt=%s, ca_crt=%s", targets, mfname, sfname, public_crt, ca_crt)
+    ret = OrderedDict()
+    ret[mfname] = verify_signature(mfname, sfname=sfname, public_crt=public_crt, ca_crt=ca_crt)
+    # ret[mfname] is the strongest claim we can make about the files we're
+    # verifiying if they match their hash in the manifest, the best we can say
+    # is whatever is the status of the manifest iteslf.
+
+    mf_dir, _ = os.path.split(mfname)
+    sf_dir, _ = os.path.split(sfname)
+
+    if mf_dir and mf_dir == sf_dir:
+        trunc = mf_dir + '/'
+    else:
+        trunc = None
+
+    # prepopulate digests with STATUS.UNKNOWN, skip things that shouldn't be
+    # digested (MANIFEST, SIGNATURE, etc) and build a database mapping
+    # normalized names back to given target names.
+    xlate = dict()
+    digests = OrderedDict()
+    if not targets:
+        targets = list(iterate_manifest(mfname))
+    for otarget in targets:
+        target = normalize_path(otarget, trunc=trunc)
+        log.debug('found manifest for %s (%s)', otarget, target)
+        if otarget != target:
+            xlate[target] = otarget
+        if target in digests or target in (mfname, sfname):
+            continue
+        digests[target] = STATUS.UNKNOWN
+    # populate digests with the hashes from the MANIFEST
+    if os.path.isfile(mfname):
+        with open(mfname, 'r') as fh:
+            for line in fh.readlines():
+                matched = MANIFEST_RE.match(line)
+                if matched:
+                    digest,manifested_fname = matched.groups()
+                    manifested_fname = normalize_path(manifested_fname)
+                    if manifested_fname in digests:
+                        digests[manifested_fname] = digest
+    # compare actual digests of files (if they exist) to the manifested digests
+    for vfname in digests:
+        digest = digests[vfname]
+        htname = os.path.join(trunc, vfname) if trunc else vfname
+        if digest == STATUS.UNKNOWN:
+            # digests[vfname] is either UNKNOWN (from the targets population)
+            # or it's a digest from the MANIFEST. If UNKNOWN, we have nothing to compare
+            # so we return UNKNOWN
+            ret[vfname] = STATUS.UNKNOWN
+        elif digest == hash_target(htname):
+            # Cool, the digest matches, but rather than mark STATUS.VERIFIED,
+            # we mark it with the same status as the MANIFEST it self --
+            # presumably it's signed (STATUS.VERIFIED); but perhaps it's only
+            # UNKNOWN or even FAIL.
+            ret[vfname] = ret[mfname]
+        else:
+            # We do have a MANIFEST entry and it doesn't match: FAIL with or
+            # without a matching SIGNATURE
+            ret[vfname] = STATUS.FAIL
+    # fix any normalized names so the caller gets back their specified targets
+    for k,v in xlate.iteritems():
```

Comment:
I think this fails in python3, would need to be `.items()`

Reply:
Yeah, I have to bring in other changes too. I'll get this tomorrow prolly.